backends/tfhe-cuda-backend/cuda/src/crypto/keyswitch.cuh

```diff
@@ -31,43 +31,68 @@ __device__ Torus *get_ith_block(Torus *ksk, int i, int level,
  *
  */
+// Each thread in x are used to calculate one output.
+// threads in y are used to paralelize the lwe_dimension_in loop.
+// shared memory is used to store intermediate results of the reduction.
 template <typename Torus>
 __global__ void
 keyswitch(Torus *lwe_array_out, Torus *lwe_output_indexes, Torus *lwe_array_in,
           Torus *lwe_input_indexes, Torus *ksk, uint32_t lwe_dimension_in,
           uint32_t lwe_dimension_out, uint32_t base_log, uint32_t level_count) {
-  int tid = threadIdx.x;
+  const int tid = threadIdx.x + blockIdx.x * blockDim.x;
+  const int shmem_index = threadIdx.x + threadIdx.y * blockDim.x;
+
   extern __shared__ int8_t sharedmem[];
+  Torus *lwe_acc_out = (Torus *)sharedmem;
+  auto block_lwe_array_out = get_chunk(
+      lwe_array_out, lwe_output_indexes[blockIdx.y], lwe_dimension_out + 1);
+
   if (tid <= lwe_dimension_out) {
-    Torus *local_lwe_array_out = (Torus *)sharedmem;
+
+    Torus local_lwe_out = 0;
     auto block_lwe_array_in = get_chunk(
-        lwe_array_in, lwe_input_indexes[blockIdx.x], lwe_dimension_in + 1);
-    auto block_lwe_array_out = get_chunk(
-        lwe_array_out, lwe_output_indexes[blockIdx.x], lwe_dimension_out + 1);
-    local_lwe_array_out[tid] = 0;
-
-    if (tid == lwe_dimension_out) {
-      local_lwe_array_out[lwe_dimension_out] =
-          block_lwe_array_in[lwe_dimension_in];
+        lwe_array_in, lwe_input_indexes[blockIdx.y], lwe_dimension_in + 1);
+
+    if (tid == lwe_dimension_out && threadIdx.y == 0) {
+      local_lwe_out = block_lwe_array_in[lwe_dimension_in];
     }
+    const Torus mask_mod_b = (1ll << base_log) - 1ll;
 
-    for (int i = 0; i < lwe_dimension_in; i++) {
+    const int pack_size = (lwe_dimension_in + blockDim.y - 1) / blockDim.y;
+    const int start_i = pack_size * threadIdx.y;
+    const int end_i = pack_size * (threadIdx.y + 1) <= lwe_dimension_in
+                          ? pack_size * (threadIdx.y + 1)
+                          : lwe_dimension_in;
+
+    // This loop distribution seems to benefit the global mem reads
+    for (int i = start_i; i < end_i; i++) {
       Torus a_i = round_to_closest_multiple(block_lwe_array_in[i], base_log,
                                             level_count);
       Torus state = a_i >> (sizeof(Torus) * 8 - base_log * level_count);
-      Torus mask_mod_b = (1ll << base_log) - 1ll;
+
       for (int j = 0; j < level_count; j++) {
         auto ksk_block =
             get_ith_block(ksk, i, j, lwe_dimension_out, level_count);
         Torus decomposed = decompose_one<Torus>(state, mask_mod_b, base_log);
-        local_lwe_array_out[tid] -= (Torus)ksk_block[tid] * decomposed;
+        local_lwe_out -= (Torus)ksk_block[tid] * decomposed;
       }
     }
-    block_lwe_array_out[tid] = local_lwe_array_out[tid];
+
+    lwe_acc_out[shmem_index] = local_lwe_out;
+  }
+
+  if (tid <= lwe_dimension_out) {
+    for (int offset = blockDim.y / 2; offset > 0 && threadIdx.y < offset;
```

Comment:
We may want to run such a reduction in other places in the future. Can you generalize and move this to its own function? That would also make this piece of code easier to read.

Reply:
This is a very specific reduction pattern, in the sense that we only want to perform the reduction on the threads in y, and only within the block. I think what you suggest could be another ticket in which we could generalise it to different configurations or different operations.

Reply:
Indeed!